[WIP] Enforce positiveness in tests using enforce_estimator_tags_X

wdevazelhes opened the pull request:
See https://github.com/scikit-learn/scikit-learn/pull/14680#issuecomment-523059108

Commit:
Enforce positiveness in tests

wdevazelhes commented:
I'll also include a test for `enforce_estimator_tags_X`. However, I noticed that `enforce_estimator_tags_y` does not have any test, should I include one too ?

Commit:
Add a test

Commit:
Add default behaviour to the test

Commit:
Add requires_positive_X to Chi2Samplers

wdevazelhes commented:
I think I'll need to change the message in the Chi2Samplers from ""Entries of X must be non-negative."" into "Negative values in data passed to", to pass `check_fit_non_negative` for those

Commit:
fix flake8

Commit:
Fix typo

wdevazelhes commented:
I'm having trouble figuring out what to do to fix the tests: the problem comes from `AdditiveChi2Sampler` and `SkewedChi2Sampler`: they should work on kernel matrices (so positive if I understand correctly), at fit and transform. However the fit actually is stateless (only uses n_features), and as such does not forbid negative `X`. Should it ? If it should not, then it means we don't need the flag `ensures_positive_X`, but then there's a problem in `check_estimators_pickle` (according to this CI report: https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=6740), because the test will not `ensure_positive_X` and therefore at the `transform` step it fails (because `X` has negative elements). Should we separate `ensure_positive_X_fit` and `ensure_positive_X_transform` ?

wdevazelhes commented:
I guess a solution is to not flag them as `requires_positive_X` (because at fit they don't strictly enforce this), but to still pass the test to to use `pairwise_estimator_convert_X` in the tests ?

wdevazelhes commented:
Ah my bad, `pairwise_estimator_convert_X` was already enforced in `check_estimators_pickle`, so there's something I don't understand

wdevazelhes commented:
Ah my bad again, turns out `SkewedChi2Sampler` and `AdditiveChi2Sampler` actually don't take kernel matrices but actual matrices of points...

wdevazelhes commented:
I'm back to the following reformulated question then:  https://github.com/scikit-learn/scikit-learn/pull/14705#issuecomment-523426310
>I'm having trouble figuring out what to do to fix the tests: the problem comes from AdditiveChi2Sampler and SkewedChi2Sampler: <strike>they should work on kernel matrices (so positive if I understand correctly), at fit and transform</strike> **EDIT: they work on positive datasets Xs.** However the fit actually is stateless (only uses n_features), and as such does not forbid negative X. Should it ? If it should not, then it means we don't need the flag ensures_positive_X, but then there's a problem in check_estimators_pickle (according to this CI report: https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=6740), because the test will not ensure_positive_X and therefore at the transform step it fails (because X has negative elements). Should we separate ensure_positive_X_fit and ensure_positive_X_transform ?

wdevazelhes commented:
However, since the Chi squared kernel works for positive points, I guess it makes sense to forbid passing negative data to `fit`, even if only `n_features` is used ?

Commit:
Ensure positiveness for chi2 samplers

wdevazelhes commented:
I just added a modification to ensure positivity at fit time too
It made me think: do you think a test `check_transform_non_negative` should also be added in addition to `check_fit_non_negative` ?

Commit:
Use the right syntax to check positiveness

Commit:
Fix pre-existing probable typo

Commit:
Remove the need for positiveness for SkewedChi2 kernel

wdevazelhes commented:
Actually I saw that `SkewedChi2Sampler` allows for negative values, so the things I said are only for `AdditiveChi2Sampler`

wdevazelhes commented:
However tests now still fail for `SkewedChi2Sampler` because `X` has values lower than a constant `c` (`SkewedChi2Sampler` takes only values higher than `c`). I'm not sure how to deal with that, should I create a new estimator tag for this case ?

amueller commented:
Hm I don't think it's worth creating a tag for that. And it's a kind of weird requirement. Not sure if anyone actually uses this and whether we should just deprecate it...

amueller commented:
Actually it's usually applied to positive data so maybe just add the tag asking for positivity, that should be fine.

Commit:
Ensure positivity for SkewedChi2Sampler

wdevazelhes commented:
> Actually it's usually applied to positive data so maybe just add the tag asking for positivity, that should be fine.

Allright, done (see commit https://github.com/scikit-learn/scikit-learn/pull/14705/commits/851887dc0de2a6f51bb588df930e9463e889b3a7)

Otherwise I was also thinking since we're talking about `X`, maybe an `X_types` estimator tag like `higher_than_c` could be used (maybe less heavy than a whole new estimator tag) ?

wdevazelhes commented on sklearn/tests/test_kernel_approximation.py:
```diff
@@ -112,11 +113,6 @@ def test_skewed_chi2_sampler():
     assert np.isfinite(kernel_approx).all(), \
         'NaNs found in the approximate Gram matrix'
 
-    # test error is raised on when inputs contains values smaller than -c
```
We should be able to remove this since now we test for positivity and it's already tested by `check_fit_non_negative` ? Although now that I think about it `check_fit_non_negative`, would only test the `fit` part, whereas the below would test the `transform` part, so I'll let it here but adapt it to check just positivity (and not > -c). 
But then it reminds me: should we have a `check_transform_non_negative` as well in common tests ?

wdevazelhes replied:
Done

amueller replied:
It's not supposed to raise an error on negative values, only on values smaller than ``-c``.

Commit:
Put back the test for positivity on transform

Commit:
Be clearer on the test

amueller commented on sklearn/kernel_approximation.py:
```diff
@@ -226,10 +227,7 @@ def transform(self, X):
-            raise ValueError("X may not contain entries smaller than"
-                             " -skewedness.")
-
+        check_non_negative(X, "SkewedChi2Sampler (input X)")
```
Why did you replace the test? Now it will fail with things that previously worked.

wdevazelhes replied:
Sorry, I thought you meant to actually change the be behavior of the estimator (and indeed make it more restrictive) here: 
>Actually it's usually applied to positive data so maybe just add the tag asking for positivity, that should be fine.

Because the thing is AFAIU I can't just add a tag `requires_positive_X` but still let the estimator accept negative values (>-c), because `check_fit_non_negative` will then fail (a test than any estimator that has a `requires_positive_X` throws an error when given negative `X`) (actually it will fail because of the error message, but even if I change the error message into something like "Negative values in data passed to" (which I could tweak into sth like "Negative values in data passed to SkewedChi2Sampler after adding the constant c", which would be more exact) the test could still fail in principle because it could in principle test only on negative values that are >-c (it does not bc the negative values are random but logically it could)

agramfort replied:
I agree with @amueller I would not change the behavior of the estimator to make the common tests pass. It just says that requires_positive_X is not what is adapted here.

wdevazelhes replied:
I agree, so should I create a new safe tag here ? Like `requires_X_higher_than_constant` or sth, that I could enforce in `enforce_estimator_tags_X` ? Because we need to find a way for tests that by default have negative values to translate them into `>c` for `SkewedChi2Sampler`

wdevazelhes replied:
Or maybe an `X_types` kind of tag ?

amueller commented on sklearn/tests/test_kernel_approximation.py:
```diff
@@ -81,16 +82,19 @@ def test_additive_chi2_sampler():
-    Y[0, 0] = -c / 2.
+
+    # approximate kernel mapping
+    transform = SkewedChi2Sampler(skewedness=c, n_components=1000,
```
It's a bad sign if you have to change any tests, because that probably means that you changed behavior and are breaking someone's code.

agramfort replied:
+1 don't do this.

wdevazelhes replied:
I agree, I forgot about those comments...  I just put the PR into WIP until there's a better solution (see comment above)

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2279,6 +2274,16 @@ def enforce_estimator_tags_y(estimator, y):
+    if _safe_tags(estimator, "requires_positive_X"):
+        # Create strictly positive X. The minimal increment above 0 is 1, as
+        # X could be of integer dtype.
+        X += 1 + abs(X.min())
```
why not ``X -= X.min() - 1e-5`` or something like that? I don't know why you would do abs. If your data had a minimum of ``2`` after this is has a minimum of ``5``

wdevazelhes replied:
I agree `X -= X.min() - 1e-5 seems better, I just had taken this line from `enforce_estimator_tags_y` above
Will do

amueller replied:
ahh the comment was about dtype int? hm not sure in how far that applies to X here. But if thests are passing then that's fine I guess ;) Is this ready? If so, please rename from WIP to MRG.

wdevazelhes replied:
Ah yes, I didn't see that comment, I guess this means if there is an integer X at one point we want to let it integer ? I don't know to what extent it applies indeed, but maybe in case we can let it with the int ? What about `X -= X.min() - 1` ? So that it'll still be ok wrt to your comment https://github.com/scikit-learn/scikit-learn/pull/14705/files/c6751b454bdf84699954e3f6bfa7000db2d42d48#r316775147 but it keeps the integer ?

Commit:
Address https://github.com/scikit-learn/scikit-learn/pull/14705#discussion_r316775147

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2279,6 +2274,16 @@ def enforce_estimator_tags_y(estimator, y):
+    # Estimators with a `requires_positive_X` tag only accept strictly positive
+    # inputs
+    if _safe_tags(estimator, "requires_positive_X"):
+        # Create strictly positive X. The minimal increment above 0 is 1, as
```
The comment and the code don't really agree, right?

wdevazelhes replied:
That's right, I forgot the comment :p (see https://github.com/scikit-learn/scikit-learn/pull/14705#discussion_r326872132)

wdevazelhes committed:
See https://github.com/scikit-learn/scikit-learn/pull/14705#discussion_r326872132

wdevazelhes committed:
Merge branch 'master' into use_requires_positive_X

wdevazelhes committed:
remove blank spaces

wdevazelhes commented:
I think if tests pass we can merge

agramfort commented:
@wdevazelhes it does not pass (yet)

wdevazelhes committed:
copy dataset to avoid inplace modification

wdevazelhes committed:
make function private

wdevazelhes committed:
Merge branch 'master' into use_requires_positive_X

wdevazelhes committed:
merge with master and update CategoricalNB too

wdevazelhes committed:
update error messages

wdevazelhes commented on sklearn/naive_bayes.py:
```diff
@@ -1168,8 +1167,7 @@ def _check_X_y(self, X, y):
         X, y = check_X_y(X, y, dtype='int')
-        if np.any(X < 0):
-            raise ValueError("X must not contain negative values.")
+        check_non_negative(X, "CategoricalNB (input X)")
```
By the way, `check_non_negative` finds the min and checks if it's negative, wouldn't it be faster to use `np.any` ? (in theory it would just need to iterate until finding one nonnegative element but I don't know if that's what is done by any)

agramfort replied:
%timeit is your friend

wdevazelhes replied:
That's right, with timeit `np.min()` is faster. I forgot but in `np.any` we still need to do the `X>0` pass, which amounts to one pass on the whole data that's why it'll not be faster... I didn't find any numpy function that would prevent to do this whole pass so I guess `np.min` is good

```python
In[1]: import numpy as np

In [2]: a = np.random.randn(10000, 10000)              
                                                                                    
In [3]: %timeit np.any(a)                                                                                                                  
124 ms ± 808 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [4]: %timeit np.min(a)                                                                                                                  
56.5 ms ± 69.8 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

wdevazelhes commented:
Tests are green now (they were failing in `test_rbf_sampler` because I updated some global dataset X and y inplace, and looks like the `rbf_sampler` didn't like the dataset updated by another test so I updated a copy)

wdevazelhes commented:
Any update on https://github.com/scikit-learn/scikit-learn/pull/14705#discussion_r329312998 ?

agramfort commented:
@wdevazelhes you need to revert the API change in SkewedChi2Sampler. Also see how to fix the merge conflicts

agramfort commented:
this is now made irrelevant after #16286
